library/src/main/java/com/heinrichreimersoftware/materialintro/app/IntroActivity.java

```diff
@@ -360,23 +360,32 @@ private long calculateScrollDuration(int distance) {
         return Math.round(pageScrollDuration * (distance + Math.sqrt(distance)) / 2);
     }
     
-    public void goToFirstSlide() {
+   public void goToFirstSlide() {
         int currentItem = pager.getCurrentItem();
-        if (currentItem > adapter.getCount() - 1) finishIfNeeded();
+        boolean canGo = false;
+        if (currentItem <= 0) return;
 
-        if (canGoForward(currentItem, true)) {
-            smoothScrollPagerTo(0);
+        while (currentItem > 0 && canGoBackward(currentItem, true)) {
+            currentItem--;
+            canGo = true;
+        }
+        if (canGo) {
```

Comment:
You should keep track of the position nearest to 0 you can move here.
Then there are three cases:

1. If you can go through all slides, just call `smoothScrollPagerTo(0);`.
2. If you can go through no slide at all, then just call `AnimUtils.applyShakeAnimation(this, buttonNext);`.`.
3. If you can go through at least one slide, then first call `smoothScrollPagerTo(nearestPositionToFirstSlide);` and after that call `AnimUtils.applyShakeAnimation(this, buttonNext);`. Of course `nearestPositionToFirstSlide` should be calculated correctly.